pkg/ingester/ingester_v2.go

```diff
@@ -58,6 +62,52 @@ type userTSDB struct {
 	ingestedRuleSamples *ewmaRate
 }
 
+// PreCreation implements SeriesLifecycleCallback interface.
+func (u *userTSDB) PreCreation(metric labels.Labels) error {
+	if u.limiter == nil {
+		return nil
+	}
+
+	// Total series limit.
+	if err := u.limiter.AssertMaxSeriesPerUser(u.userID, int(u.DB.Head().NumSeries())); err != nil {
+		return makeLimitError(perUserSeriesLimit, err)
+	}
+
+	// Series per metric name limit.
+	metricName, err := extract.MetricNameFromLabels(metric)
+	if err != nil {
+		// TODO(codesome): what should be the action?
+		return nil
+	}
+	if err := u.seriesInMetric.canAddSeriesFor(u.userID, metricName); err != nil {
+		return makeMetricLimitError(perMetricSeriesLimit, metric, err)
+	}
+
+	return nil
+}
+
+// PostCreation implements SeriesLifecycleCallback interface.
+func (u *userTSDB) PostCreation(metric labels.Labels) {
+	metricName, err := extract.MetricNameFromLabels(metric)
+	if err != nil {
+		// TODO(codesome): what should be the action?
+		return
+	}
+	u.seriesInMetric.seriesAddedFor(metricName)
+}
+
+// PostDeletion implements SeriesLifecycleCallback interface.
+func (u *userTSDB) PostDeletion(metrics ...labels.Labels) {
+	for _, metric := range metrics {
+		metricName, err := extract.MetricNameFromLabels(metric)
+		if err != nil {
+			// TODO(codesome): what should be the action?
```

Comment:
I would just replace this TODO with a comment saying:
`This should never happen because metrics have already been checked in PreCreation().`